### DISCUSSION TITLE
Driver Breaks

### DISCUSSION THREAD

**Discussion by GuyBenhaim** [google/or-tools#2982] [Routing (and legacy CP) questions]
Hello,
For Google.OrTools.7.5.7466 in .Net, using Routing

I'm checking Breaks using IntervalVar. Aiming for one break per vehicle, that occurs anytime within a certain time-window.
With a small set (1 vehicle, 2 nodes: P + D) adding the breaks does not seem to change the solution times at all. I do get a solution.

I used the newer version of the API:
```cpp
SetBreakIntervalsOfVehicle(
    std::vector<IntervalVar*> breaks,
    int vehicle,
    int pre_travel_evaluator,
    int post_travel_evaluator);
```
To simplify I set the last two parameters to -1 (which should make these return 0 for pre and post times).

Anyone successful with implementing such case?  (i.e. getting a correct and consistent solution)
Can this API work, are we forced to use wacky fake-nodes?

p.s. Posted also in GitHub

Thanks

**Comment by Mizux** [comment]
did you try master ? v7.5 is quite old and IIRC there was some dev on break recently...

**Comment by GuyBenhaim** [comment]
Changed to v7.8 and used this:
```cpp
break_intervals[vehicle] = model.solver().MakeFixedDurationIntervalVar(
                             0, end_time, 10000, false, "vehicle " + vehicle.ToString());
time_dimension.SetBreakIntervalsOfVehicle(break_intervals, vehicle, -1,-1);
```

With the SetBreakIntervalsOfVehicle condition solver seems to push the visits to be **as late as possible**, i.e. the solution always visits the end point at the latest possibe (end time = 60000). This occurs even when setting IntervalVar Duration=0.
However, commenting-out just this condition pulls the visits to be **as early as possible** and the end point is visited early.
The total time remains the same, the differences in visit times are the same ....

For example if visits with the condition are 57000 > 58000 > 59000 > 60000
Voiding it gives 0 > 1000 > 2000 > 3000

**Comment by GuyBenhaim** [comment]
Probing IntervalVar:
long startvalue = solution.StartValue(brk_var);
long endvalue = solution.EndValue(brk_var);

It is **performed and gets start and end values assigned to it**, but only pushes the visit times (i.e. no effect on time between node and total solution time )
Its gets more weird: setting span cost to 0, prevents the visits being pushed to the latest possible, but **still no effect on visits and total time**.
time_dimension.SetSpanCostCoefficientForVehicle(0, vehicle);

**Comment by GuyBenhaim** [comment]
I would like to know whether a basic **one-break with one vehicle** case works using  time_dimension.SetBreakIntervalsOfVehicle 
Any experience ? I did not yet mange to get this to work ... (v7.5 or v7.8)

**Comment by GuyBenhaim** [comment]
Is there a way for the vehicle break to start certain T time-units after the vehicle's start time, which is 
IntVar **start** = time_dimension.CumulVar(routing.Start(v))

Either by setting the intervalvar's  min_start and max_start parameters, using **start**:
break_intervals[vehicle] = model.solver().MakeFixedDurationIntervalVar(**start**, **start**+T, 10000, false, "vehicle " + vehicle.ToString());

Or by adding a proper constraint:
model.solver().MakeIntervalVarRelation(break_interval, **start**) ?
model.solver().Add(break_interval.StartMin() > **start**) ?
break_interval.SetStartMin(**start**) ?

**Comment by GuyBenhaim** [comment]
Any ideas?  When vehicles start after t=0, defining breaks based on an absolute start time does not work ...

**Comment by GuyBenhaim** [comment]
Is this a feature to be supported? or a possible to trick to support on 7.X ?

**Comment by Mizux** [comment]
so using a duration dimension like explaining in another thread won't make it ?

**Comment by GuyBenhaim** [comment]
Should work as long as no is added Slack in the time-dimension.

BTW: I saw Laurent's earlier request. Should we move all discussion to GitHub?

**Comment by lperron** [comment]
move: If you want to. I will keep both systems in parallel to
understand trends and preferences.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by Mizux** [comment]
Since break is a hot and specific topic, I like github issue since you can have several labels while discussion seems to be one category only. 😞 
e.g.: https://github.com/google/or-tools/labels/routing_break it's nice to see all related routing break issues/topics
note: waiting for https://github.com/github/feedback/discussions/19 🤞
